File: ios/Classes/ArcgisFlutterPlugin.m
#import "ArcgisFlutterPlugin.h"
#import "MapViewController.h"
#import <ArcGIS/ArcGIS.h>
#import "MapAnnotation.h"
#import "MapLayer.h"

@implementation ArcgisFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"com.apptreesoftware.arcgis_flutter_plugin"
            binaryMessenger:[registrar messenger]];
    UIViewController *host = UIApplication.sharedApplication.delegate.window.rootViewController;
    ArcgisFlutterPlugin* instance = [[ArcgisFlutterPlugin alloc] initWithHost:host channel:channel registrar:registrar];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)initWithHost:(UIViewController *)host channel:(FlutterMethodChannel *)channel registrar:(NSObject<FlutterPluginRegistrar>*)registar {
    if (self = [super init]) {
        self.host = host;
        self.channel = channel;
        self.redImageKey = [registar lookupKeyForAsset:@"packages/arcgis_flutter/lib/assets/pin_red.png"];
        self.layerImageKey = [registar lookupKeyForAsset:@"packages/arcgis_flutter/lib/assets/layers.png"];
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"setApiKey" isEqualToString:call.method]) {
        NSString *licenseStr = (NSString*) call.arguments;
        if (licenseStr.length == 0) {
            result(@YES);
            return;
        }
        [AGSArcGISRuntimeEnvironment setLicenseKey:licenseStr error:nil];
        result(@YES);
    } else if ([@"show" isEqualToString:call.method]) {
        NSDictionary *args = call.arguments;
        NSDictionary *mapOptions = args[@"mapOptions"];
        
        MapViewController *vc = [[MapViewController alloc] initWithPlugin:self
                                 navigationItems:[self buttonItemsFromActions:args[@"actions"]]
                                 ];
        UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:vc];
        navController.navigationBar.translucent = NO;
        [self.host presentViewController:navController animated:true completion:nil];
        self.mapViewController = vc;
        result(@YES);
    } else if ([@"dismiss" isEqualToString:call.method]) {
        if (self.mapViewController) {
            [self.host dismissViewControllerAnimated:true completion:nil];
        }
        self.mapViewController = nil;
    } else if ([@"setAnnotations" isEqualToString:call.method]) {
        [self handleSetAnnotations:call.arguments];
        result(@YES);
    } else if ([@"addAnnotation" isEqualToString:call.method]) {
        [self handleAddAnnotation:call.arguments];
        result(@YES);
    } else if ([@"setCamera" isEqualToString:call.method]) {
        [self handleSetCamera:call.arguments];
        result(@YES);
    } else if ([@"getCenter" isEqualToString:call.method]) {
        CLLocationCoordinate2D location = [self.mapViewController centerLocation];
        result(@{@"latitude": @(location.latitude), @"longitude": @(location.longitude)});
    } else if ([@"getZoomLevel" isEqualToString:call.method]) {
        result(@([self.mapViewController zoomLevel]));
    } else if ([@"setLayers" isEqualToString:call.method]) {
        [self handleAddLayers:call.arguments];
        result(@YES);
        
    } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleSetCamera:(NSDictionary *)cameraUpdate {
    CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake([cameraUpdate[@"latitude"] doubleValue], [cameraUpdate[@"longitude"] doubleValue]);
    [self.mapViewController setCamera:coordinate zoom:[cameraUpdate[@"zoom"] floatValue]];
}

- (void)onMapReady {
    [self.channel invokeMethod:@"onMapReady" arguments:nil];
}

- (void)handleAddAnnotation:(NSDictionary *)dict {
    MapAnnotation *annotation = [MapAnnotation annotationFromDictionary:dict];
    if (annotation) {
        [self.mapViewController addAnnotation:annotation];
    }
}

-(void)handleSetAnnotations:(NSDictionary *)annotations {
    NSMutableArray *array = [NSMutableArray array];
    for (NSDictionary *aDict in annotations) {
        MapAnnotation *annotation = [MapAnnotation annotationFromDictionary:aDict];
        if (annotation) {
            [array addObject:annotation];
        }
    }
    [self.mapViewController updateAnnotations:array];
}

- (NSArray *)buttonItemsFromActions:(NSArray *)actions {
    NSMutableArray *buttons = [NSMutableArray array];
    if (actions) {
        for (NSDictionary *action in actions) {
            UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithTitle:[action valueForKey:@"title"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(handleToolbar:)];
            button.tag = [[action valueForKey:@"identifier"] intValue];
            [buttons addObject:button];
        }
    }
    return buttons;
}

- (void)handleToolbar:(UIBarButtonItem *)item {
    [self.channel invokeMethod:@"onToolbarAction" arguments:@(item.tag)];
}

- (void)handleAddLayers:(NSDictionary *)dict {
    NSMutableArray *typedLayers = [[NSMutableArray alloc] init];
    NSArray *rawLayers = (NSArray *)[dict objectForKey:@"mapLayers"];
    for (int i = 0; i < [rawLayers count];i++) {
        MapLayer *layer = [[MapLayer alloc] initWithDictionary:rawLayers[i]];
        [typedLayers addObject:layer];
    }
    [self.mapViewController setLayers:typedLayers];
}

@end

File: android/src/main/kotlin/com/apptreesoftware/arcgisflutter/ArcgisFlutterPlugin.kt
package com.apptreesoftware.arcgisflutter

import android.app.Activity
import android.content.Intent
import com.esri.arcgisruntime.ArcGISRuntimeEnvironment
import com.esri.arcgisruntime.geometry.Point
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar


class ArcgisFlutterPlugin(val activity: Activity) : MethodCallHandler {
  companion object {
    lateinit var channel: MethodChannel
    var toolbarActions: List<ToolbarAction> = emptyList()
    var showUserLocation: Boolean = false
    var mapActivity: MapActivity? = null

    @JvmStatic
    fun registerWith(registrar: Registrar): Unit {
      channel = MethodChannel(registrar.messenger(), "com.apptreesoftware.arcgis_flutter_plugin")
      val plugin = ArcgisFlutterPlugin(activity = registrar.activity())
      channel.setMethodCallHandler(plugin)
    }

    fun handleToolbarAction(id: Int) {
      channel.invokeMethod("onToolbarAction", id)
    }

    fun onMapReady() {
      channel.invokeMethod("onMapReady", null)
    }

    fun getToolbarActions(actionsList: List<Map<String, Any>>?): List<ToolbarAction> {
      if (actionsList == null) return emptyList()
      val actions = ArrayList<ToolbarAction>()
      actionsList.mapTo(actions) { ToolbarAction(it) }
      return actions
    }
  }


  override fun onMethodCall(call: MethodCall, result: Result): Unit {
    when {
      call.method == "setApiKey" -> {
        val licenseStr = call.arguments as String
        if (licenseStr.isEmpty()) {
          result.success(true)
        }
        ArcGISRuntimeEnvironment.setLicense(licenseStr)
        result.success(true)
      }
      call.method == "show" -> {
        val mapOptions = call.argument<Map<String, Any>>("mapOptions")
        toolbarActions = getToolbarActions(call.argument<List<Map<String, Any>>>("actions"))
        showUserLocation = mapOptions["showUserLocation"] as Boolean

        val intent = Intent(activity, MapActivity::class.java)
        activity.startActivity(intent)
        result.success(true)
        return
      }
      call.method == "dismiss" -> {
        mapActivity?.finish()
        result.success(true)
        return
      }
      call.method == "setLayers" -> {
        setLayers(call.arguments as Map<String, Any>)
        result.success(true)
      }
      call.method == "getZoomLevel" -> {
        val zoom = mapActivity?.zoom ?: 0.0.toFloat()
        result.success(zoom)
      }
      call.method == "getCenter" -> {
        val center = mapActivity?.target ?: Point(0.0, 0.0)
        result.success(mapOf("latitude" to center.y,
                "longitude" to center.x))
      }
      call.method == "setCamera" -> {
        handleSetCamera(call.arguments as Map<String, Any>)
        result.success(true)
      }

      call.method == "setAnnotations" -> {
        handleSetAnnotations(call.arguments as List<Map<String, Any>>)
        result.success(true)
      }
      call.method == "addAnnotation" -> {
        handleAddAnnotation(call.arguments as Map<String, Any>)
      }
      else -> result.notImplemented()
    }
  }

  fun handleSetCamera(map: Map<String, Any>) {
    val lat = map["latitude"] as Double
    val lng = map["longitude"] as Double
    val zoom = map["zoom"] as Double
    mapActivity?.setCamera(LatLng(lat, lng), zoom.toFloat())
  }
  fun setLayers(map: Map<String, Any>) {
    var layers = map["mapLayers"] as List<Map<String, Any>>
    var typedLayers = layers.map { MapLayer.fromMap(it) }
    typedLayers = typedLayers.filterNotNull()
    mapActivity?.setLayers(typedLayers)
  }

  fun handleSetAnnotations(annotations: List<Map<String, Any>>) {
    val mapAnnoations = ArrayList<MapAnnotation>()
    for (a in annotations) {
      val mapAnnotation = MapAnnotation.fromMap(a)
      if (mapAnnotation != null) {
        mapAnnoations.add(mapAnnotation)
      }
    }
    mapActivity?.setAnnotations(mapAnnoations)
  }

  fun handleAddAnnotation(map: Map<String, Any>) {
    MapAnnotation.fromMap(map)?.let {
      mapActivity?.addMarker(it)
    }
  }

}